althea_types/src/websockets/encryption.rs

```diff
@@ -0,0 +1,221 @@
+                write!(f, "UTF-8 Conversion error: {}", e)
+            }
+        }
+    }
+}
+
+pub fn encrypt_router_websocket_msg(
+    our_publickey: WgKey,
+    our_secretkey: &SecretKey,
+    ops_pubkey: &PublicKey,
+    message: RouterWebsocketMessage,
+) -> EncryptedRouterWebsocketMessage {
+    let plaintext = serde_json::to_string(&message)
+        .expect("Failed to serialize router websocket message")
+        .into_bytes();
+    let nonce = SalsaBox::generate_nonce(&mut OsRng);
+    let b = SalsaBox::new(ops_pubkey, our_secretkey);
+    let ciphertext = b.encrypt(&nonce, plaintext.as_ref()).unwrap();
+    EncryptedRouterWebsocketMessage {
+        pubkey: our_publickey,
+        nonce: nonce.into(),
+        encrypted_router_websocket_msg: ciphertext,
+    }
+}
+
+pub fn decrypt_router_websocket_msg(
+    our_secretkey: &SecretKey,
+    message: EncryptedRouterWebsocketMessage,
+) -> Result<RouterWebsocketMessage, WebsocketEncryptionError> {
+    let their_nacl_pubkey = PublicKey::from(message.pubkey);
+    let their_nonce = message.nonce;
+    let ciphertext = message.encrypted_router_websocket_msg;
+    let b = SalsaBox::new(&their_nacl_pubkey, our_secretkey);
+    let decrypted_bytes = match b.decrypt(their_nonce.as_ref().into(), ciphertext.as_ref()) {
+        Ok(value) => value,
+        Err(_) => {
+            return Err(WebsocketEncryptionError::DecryptionError {
+                e: "Could not decrypt websocket message".to_string(),
+            })
+        }
+    };
+    let decrypted_string = match String::from_utf8(decrypted_bytes) {
+        Ok(value) => value,
+        Err(_) => {
+            return Err(WebsocketEncryptionError::Utf8Error {
+                e: "Could not convert decrypted bytes to string".to_string(),
+            })
+        }
+    };
+    let decrypted_message = match serde_json::from_str(&decrypted_string) {
+        Ok(value) => value,
+        Err(_) => {
+            return Err(WebsocketEncryptionError::DecryptionError {
+                e: "Could not deserialize decrypted string".to_string(),
+            })
+        }
+    };
+
+    Ok(decrypted_message)
+}
+
+/// This message goes from ops to the router
+pub fn encrypt_ops_websocket_msg(
+    our_publickey: WgKey,
+    our_secretkey: &SecretKey,
+    their_publickey: &PublicKey,
+    message: OperatorWebsocketMessage,
+) -> EncryptedOpsWebsocketMessage {
+    let plaintext = serde_json::to_string(&message)
+        .expect("Failed to serialize operator websocket message")
+        .into_bytes();
+    let nonce = SalsaBox::generate_nonce(&mut OsRng);
+    let b = SalsaBox::new(their_publickey, our_secretkey);
+    let ciphertext = b.encrypt(&nonce, plaintext.as_ref()).unwrap();
+    EncryptedOpsWebsocketMessage {
+        nonce: nonce.into(),
+        encrypted_ops_websocket_msg: ciphertext,
+        pubkey: our_publickey,
+    }
+}
+
+pub fn decrypt_ops_websocket_msg(
+    our_secretkey: &SecretKey,
+    ops_publickey: &PublicKey,
```

Comment:
implementing these encrypt and decrypt as member functions of the struct might be useful so you can go

encryptedopsmsg.decrypt(args)

opsmsg.encrypt(args)

You don't need to move the functions around to do so, just wrap them in an impl Optoolsmessage or impl EncryptedOpsMessage for example.